hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/impl/MutableStateBase.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.state.impl;
+
+import com.hedera.node.app.spi.state.WritableState;
+import com.swirlds.common.merkle.MerkleNode;
+import java.util.*;
+import javax.annotation.Nonnull;
+
+/**
+ * A base class for implementations of {@link WritableState}.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+abstract class MutableStateBase<K, V> extends StateBase<K, V> implements WritableState<K, V> {
+    /** A map of all modified values buffered in this mutable state */
+    private final Map<K, V> modified = new HashMap<>();
+
+    /** A set of all keys (and their implicit values) removed by this mutable state */
+    private final Set<K> removed = new HashSet<>();
+
+    /**
+     * Create a new StateBase.
+     *
+     * @param stateKey The state key. Cannot be null.
+     */
+    MutableStateBase(@Nonnull final String stateKey) {
+        super(stateKey);
+    }
+
+    /** Flushes all changes into the underlying data store. */
+    public final void commit() {
+        for (final K key : removed) {
+            removeFromDataSource(key);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Clears the set of modified keys and removed keys. Equivalent semantically to a "rollback"
+     * operation.
+     */
+    @Override
+    public final void reset() {
+        super.reset();
```

Comment:
If any `MutableStateBase` subclass will have a singleton lifecycle, we will need a way to order its keys so that reconnected nodes will get the same iteration order over `removed` and `modified` in `commit()`.

Reply:
We don't have a singleton use case (even a rebuilt data structure won't be singleton -- the underlying data structure will be, but the state is a wrapper that buffers changes and records accesses. We wouldn't want that to be immutable).

Still, I am not getting the danger even if we did have singleton lifecycle, because I don't think State will be part of reconnect. I'm probably missing something very important in my mental model!